apis/v1alpha1/httproute_types.go

```diff
@@ -326,6 +343,38 @@ type HTTPHeaderMatch struct {
 	Values map[string]string `json:"values"`
 }
 
+// HTTPQueryParamMatch describes how to select a HTTP route by matching HTTP
+// query parameters.
+type HTTPQueryParamMatch struct {
+	// Type specifies how to match against the value of the query parameter.
+	//
+	// Support: Extended (Exact)
+	//
+	// Support: Custom (RegularExpression, ImplementationSpecific)
+	//
+	// Since RegularExpression QueryParamMatchType has custom conformance,
+	// implementations can support POSIX, PCRE or any other dialects of regular
+	// expressions. Please read the implementation's documentation to determine
+	// the supported dialect.
+	//
+	// HTTP query parameter matching MUST be case-sensitive for both keys and
+	// values.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *QueryParamMatchType `json:"type,omitempty"`
+
+	// Values is a map of HTTP query parameters to be matched. It MUST contain
+	// at least one entry.
+	//
+	// The query parameter name to match is the map key, and the value of the
+	// query parameter is the map value.
+	//
+	// Multiple match values are ANDed together, meaning, a request must match
+	// all the specified query parameters to select the route.
```

Comment:
Nit: I feel like we should be clear that, even if you specify one of the ImplementationSpecific matches, we shouldn't allow regexes or other craziness in the query param *name*, just the value?

Something like:
```suggestion
	// all the specified query parameters to select the route.
	//
	// Note that the query param type matching MUST only be used for the query param value, not the query param name.
```

Reply:
More direct would be (don't refer to param matching type to avoid indirection):
```
Query parameter key is must be an exact match (string comparison).
```

Reply:
Good call, added in #636